### PR TITLE
Stats: Only show tiers higher than site usage

### PR DESCRIPTION
**Pull request by kangzj**
Related: 
- https://github.com/Automattic/wp-calypso/issues/85246
- p1712791967808509/1712791561.443719-slack-C82FZ5T4G

## Proposed Changes

* Existing users will see the tier that matches their site traffic to start with
* Always show six tiers to users, not only when we run out of tiers
* Removed some obsolete code

## Testing Instructions

* Open Calypso Live for a site with 10k+ views `/stats/purchase/:siteSlug`
* Ensure the tier starts from `100k` or higher - you could try jetpack.com
* Open Calypso Live for a site with <10k views `/stats/purchase/:siteSlug`
* Open WordPress.com for the same site
* Ensure the tiers and prices match
* Purchase `100k` tier
* Open Calypso Live for the site again `/stats/purchase/:siteSlug`
* Ensure you see six tiers with correct prices

image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/my-sites/stats/stats-purchase/test/test-use-available-upgrade-tiers.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+		const usageData = { views_limit: null, billableMonthlyViews: 0 };
+		const extendedTiers = extendTiersBeyondHighestTier( tiers, CURRENCY_CODE, usageData );
+
+		expect( extendedTiers.length ).toBe( 6 );
+		expect( extendedTiers[ 0 ] ).toEqual( tiers[ 0 ] );
+		expect( extendedTiers[ 1 ] ).toEqual( tiers[ 1 ] );
+	} );
+	it( 'should return original tiers if purchased', () => {
+		const usageData = { views_limit: 10000, billableMonthlyViews: 0 };
+		const newTiers = filterLowerTiers( tiers, usageData );
```
Comment (kangzj) [inline client/my-sites/stats/stats-purchase/test/test-use-available-upgrade-tiers.js]: we should test those functions separately. Or just test the top level function. Will do tomorrow.

File: client/my-sites/stats/hooks/use-site-complusory-plan-selection-qualified-check.ts [hunk 2]
```diff
@@ -9,17 +9,11 @@ export default function useSiteComplusoryPlanSelectionQualifiedCheck( siteId: nu
...
-		usageInfo?.recent_usages[ 1 ]?.views_count ?? 0,
-		usageInfo?.recent_usages[ 2 ]?.views_count ?? 0,
-		usageInfo?.current_usage?.views_count ?? 0
-	);
+
 	const isNewSite =
 		siteCreatedTimeStamp && new Date( siteCreatedTimeStamp ) > new Date( '2024-01-31' ); // Targeting new sites
-	const isExistingSampleSite = recentMonthlyViews > MIN_MONTHLY_VIEWS_TO_APPLY_PAYWALL; // Targeting all existing sites with views higher than 1000/mth.
+	const isExistingSampleSite =
+		usageInfo?.billableMonthlyViews ?? 0 > MIN_MONTHLY_VIEWS_TO_APPLY_PAYWALL; // Targeting all existing sites with views higher than 1000/mth.
```
Comment (grzegorz-cp) [inline client/my-sites/stats/hooks/use-site-complusory-plan-selection-qualified-check.ts]: This looks like a typo with `0`.

Comment (kangzj) [reply client/my-sites/stats/hooks/use-site-complusory-plan-selection-qualified-check.ts]: Oh this is intended - mostly for the sake of typing

Comment (kangzj) [reply client/my-sites/stats/hooks/use-site-complusory-plan-selection-qualified-check.ts]: Thanks for catching this! Addressed in f74dbefef06

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx [hunk 3]
```diff
@@ -107,43 +115,22 @@ function extendTiersBeyondHighestTier(
 				( highestTier?.views ?? 0 ) +
 				( highestTier.transform_quantity_divide_by ?? 0 ) * extendedTierCount;
 
-			const extendedTierAmount = extendedTierCount - purchasedExtendedTierCount;
-
 			availableTiers.push( {
 				minimum_price: totalPrice,
-				// Upgrade price for every 1M views.
-				upgrade_price: ( highestTier.per_unit_fee || 0 ) * extendedTierAmount,
+				upgrade_price: totalPrice - usageData?.current_tier?.minimum_price ?? 0,
```
Comment (dognose24) [inline client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: The `usageData?current_tier.minimum_price` would be limited to the `2M` tier price when users have purchased an extended tier higher than `2M`. We should calculate it with `usageData?.views_limit`.

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx [hunk 4]
```diff
@@ -83,14 +88,17 @@ function extendTiersBeyondHighestTier(
...
-	if ( availableTiers.length === 1 && !! availableTiers[ 0 ].transform_quantity_divide_by ) {
-		const highestTier = availableTiers[ 0 ];
-
-		const purchasedExtendedTierCount =
-			usageData?.views_limit / ( highestTier.transform_quantity_divide_by || 1 ) -
-			EXTENSION_THRESHOLD_IN_MILLION;
-		const extendedTierCountStart = purchasedExtendedTierCount + 1;
-		const extendedTierCountEnd = extendedTierCountStart + EXTENDED_TIERS_AMOUNT;
+	const highestTier = availableTiers[ availableTiers.length - 1 ];
+	// Remove the first tier, which is used to extend higher tiers.
```
Comment (dognose24) [inline client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: When I tested with the purchased `1M` tier, the starting extended tier would be the base calculating tier, which should be omitted.

截圖 2024-04-13 上午12 27 08

https://github.com/Automattic/wp-calypso/assets/6869813/833074b7-2b5f-4ac0-b46d-250bd7515a59

**Review by dognose24 — CHANGES_REQUESTED** [review]
@kangzj Thanks for the complex calculation work! I have tested it by manually updating the `views_count` of `recent_usages` and found some issues on extended tiers.

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx (hunk 4 above)
Comment (dognose24) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: Should be addressed in https://github.com/Automattic/wp-calypso/pull/89429/commits/c9b8b3e03c172b9c4b6569d5cb0b58c27a229905.

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx (hunk 3 above)
Comment (dognose24) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: Should be addressed in https://github.com/Automattic/wp-calypso/pull/89429/commits/d5b8901d1814acfdf8f42761cce3da829f3d42e8.

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx [hunk 5]
```diff
@@ -83,14 +88,34 @@ function extendTiersBeyondHighestTier(
 	currencyCode: string,
 	usageData: PlanUsage
 ): StatsPlanTierUI[] {
-	if ( availableTiers.length === 1 && !! availableTiers[ 0 ].transform_quantity_divide_by ) {
-		const highestTier = availableTiers[ 0 ];
+	const highestTier = availableTiers[ availableTiers.length - 1 ];
+	if ( availableTiers.length < MAX_TIERS_NUMBER && !! highestTier.transform_quantity_divide_by ) {
+		// Calculate the number of tiers to extend based on either current purchase or billable monthly views.
+		const startingTier =
```
Comment (dognose24) [inline client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx:94]: The tiers would not be integers if the `billableMonthlyViews` is more than 2 million and not divisible by 1 million.

截圖 2024-04-13 上午3 51 38

Comment (dognose24) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx:94]: Addressed in https://github.com/Automattic/wp-calypso/pull/89429/commits/10726e69e3736770c68358f5bfd5f45f5b98d2e2.

**Review by dognose24 — APPROVED** [review]
It works as described after addressing some issues with the extended tiers.

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx (hunk 3 above)
Comment (kangzj) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: Added test for it: d27851d9688

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx (hunk 4 above)
Comment (kangzj) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx]: Added test to cover this: 66605a4ab0c

File: client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx (hunk 5 above)
Comment (kangzj) [reply client/my-sites/stats/hooks/use-available-upgrade-tiers.tsx:94]: Test covered in 334a97622b3

**Comment by kangzj** [comment]
@dognose24 Thank you so much for testing and fixing it for different scenarios ❤️  I added some tests to cover those cases. Let's ship it away 🎉

**Pull request merged**